### PR TITLE
En 3583 add is interface nil on all interfaces

### PR DESCRIPTION
**Pull request by BeniaminDrasovean**
Add IsInterfaceNil method on all interfaces and implementations.

### CODE REVIEW

File: process/smartContract/process.go [hunk 1]
```diff
@@ -4,10 +4,11 @@ import (
 	"bytes"
 	"encoding/hex"
 	"fmt"
-	"github.com/ElrondNetwork/elrond-go/process/factory"
 	"math/big"
 	"sync"
 
+	"github.com/ElrondNetwork/elrond-go/process/factory"
+
```
Comment (sasurobert) [inline process/smartContract/process.go]: goimports

File: process/sync/shardblock_test.go [hunk 2]
```diff
@@ -945,7 +945,7 @@ func TestBootstrap_SyncBlockShouldCallForkChoice(t *testing.T) {
 	shardCoordinator := mock.NewOneShardCoordinatorMock()
 	account := &mock.AccountsStub{}
 
-	rnd, _ := round.NewRound(time.Now(), time.Now(), time.Duration(100*time.Millisecond), mock.SyncTimerMock{})
+	rnd, _ := round.NewRound(time.Now(), time.Now(), time.Duration(100*time.Millisecond), &mock.SyncTimerMock{})
```
Comment (sasurobert) [inline process/sync/shardblock_test.go]: are you sure ? mock might implement directly, without pointer. In that case it works both with and without pointer.

File: cmd/node/factory/structs.go [hunk 3]
```diff
@@ -549,6 +549,14 @@ func (*nullChronologyValidator) ValidateReceivedBlock(shardID uint32, epoch uint
 	return nil
 }
 
+// IsInterfaceNil return if there is no value under the interface
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply cmd/node/factory/structs.go]: Done.

File: consensus/mock/blockChainMock.go [hunk 4]
```diff
@@ -146,3 +146,12 @@ func (bc *BlockChainMock) PutBadBlock(blockHash []byte) {
 		bc.PutBadBlockCalled(blockHash)
 	}
 }
+
+// IsInterfaceNil returns if there is no value under the interface
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply consensus/mock/blockChainMock.go]: done.

File: consensus/mock/blocksTrackerMock.go [hunk 5]
```diff
@@ -31,3 +31,11 @@ func (btm *BlocksTrackerMock) SetBlockBroadcastRound(nonce uint64, round int64)
 func (btm *BlocksTrackerMock) BlockBroadcastRound(nonce uint64) int64 {
 	return btm.BlockBroadcastRoundCalled(nonce)
 }
+
+// IsInterfaceNil return if there is no value under the interface
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply consensus/mock/blocksTrackerMock.go]: done

File: consensus/mock/bootstrapMock.go [hunk 6]
```diff
@@ -47,3 +47,11 @@ func (boot *BootstrapperMock) StartSync() {
 func (boot *BootstrapperMock) StopSync() {
 	boot.StopSyncCalled()
 }
+
+// IsInterfaceNil return if there is no value under the interface
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply consensus/mock/bootstrapMock.go]: done

File: consensus/mock/forkDetectorMock.go [hunk 7]
```diff
@@ -37,3 +37,11 @@ func (fdm *ForkDetectorMock) ProbableHighestNonce() uint64 {
 func (fdm *ForkDetectorMock) ResetProbableHighestNonceIfNeeded() {
 	fdm.ResetProbableHighestNonceIfNeededCalled()
 }
+
+// IsInterfaceNil return if there is no value under the interface
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply consensus/mock/forkDetectorMock.go]: done

File: consensus/mock/rounderMock.go [hunk 8]
```diff
@@ -54,3 +54,11 @@ func (rndm *RounderMock) RemainingTime(startTime time.Time, maxTime time.Duratio
 
 	return time.Duration(4000 * time.Millisecond)
 }
+
+// IsInterfaceNil returns if there is no value under the interface
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply consensus/mock/rounderMock.go]: done

File: consensus/mock/shardCoordinatorMock.go [hunk 9]
```diff
@@ -35,3 +35,11 @@ func (scm ShardCoordinatorMock) CommunicationIdentifier(destShardID uint32) stri
 
 	return "_0"
 }
+
+// IsInterfaceNil return if there is no value under the interface
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply consensus/mock/shardCoordinatorMock.go]: done

File: consensus/mock/singleSigMock.go [hunk 10]
```diff
@@ -16,3 +16,12 @@ func (s *SingleSignerMock) Sign(private crypto.PrivateKey, msg []byte) ([]byte,
 func (s *SingleSignerMock) Verify(public crypto.PublicKey, msg []byte, sig []byte) error {
 	return s.VerifyStub(public, msg, sig)
 }
+
+// IsInterfaceNil returns if there is no value under the interface
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply consensus/mock/singleSigMock.go]: done

File: consensus/mock/sposWorkerMock.go [hunk 11]
```diff
@@ -56,3 +56,12 @@ func (sposWorkerMock *SposWorkerMock) ExecuteStoredMessages() {
 func (sposWorkerMock *SposWorkerMock) BroadcastUnnotarisedBlocks() {
 	sposWorkerMock.BroadcastUnnotarisedBlocksCalled()
 }
+
+// IsInterfaceNil returns if there is no value under the interface
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply consensus/mock/sposWorkerMock.go]: done

File: consensus/mock/validatorGroupSelectorMock.go [hunk 12]
```diff
@@ -45,3 +45,8 @@ func (vgsm ValidatorGroupSelectorMock) SetConsensusGroupSize(int) error {
 func (vgsm ValidatorGroupSelectorMock) GetSelectedPublicKeys(selection []byte) (publicKeys []string, err error) {
 	panic("implement me")
 }
+
+// IsInterfaceNil returns if there is no value under the interface
+func (vgsm ValidatorGroupSelectorMock) IsInterfaceNil() bool {
+	panic("implement  me")
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply consensus/mock/validatorGroupSelectorMock.go]: done

File: consensus/mock/validatorMock.go [hunk 13]
```diff
@@ -25,3 +25,11 @@ func (vm *ValidatorMock) Rating() int32 {
 func (vm *ValidatorMock) PubKey() []byte {
 	return vm.pubKey
 }
+
+// IsInterfaceNil returns if there is no value under the interface
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply consensus/mock/validatorMock.go]: done.

File: consensus/spos/bn/bnSubroundsFactory.go [hunk 14]
```diff
@@ -369,3 +369,12 @@ func (fct *factory) initConsensusThreshold() {
 	fct.consensusState.SetThreshold(SrCommitment, pbftThreshold)
 	fct.consensusState.SetThreshold(SrSignature, pbftThreshold)
 }
+
+// IsInterfaceNil returns if there is no value under the interface
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply consensus/spos/bn/bnSubroundsFactory.go]: done

File: consensus/spos/bn/bnWorker.go [hunk 15]
```diff
@@ -71,3 +71,12 @@ func (wrk *worker) CanProceed(consensusState *spos.ConsensusState, msgType conse
 func (wrk *worker) IsMessageWithBlockHeader(msgType consensus.MessageType) bool {
 	return msgType == MtBlockHeader
 }
+
+// IsInterfaceNil returns if there is no value under the interface
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply consensus/spos/bn/bnWorker.go]: done

File: consensus/spos/consensusCore.go [hunk 16]
```diff
@@ -149,3 +149,12 @@ func (cc *ConsensusCore) RandomnessPrivateKey() crypto.PrivateKey {
 func (cc *ConsensusCore) RandomnessSingleSigner() crypto.SingleSigner {
 	return cc.blsSingleSigner
 }
+
+// IsInterfaceNil returns if there is no value under the interface
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply consensus/spos/consensusCore.go]: done

File: data/mock/shardCoordinatorMock.go [hunk 17]
```diff
@@ -37,3 +37,11 @@ func (scm ShardCoordinatorMock) CommunicationIdentifier(destShardID uint32) stri
 
 	return "_0"
 }
+
+// IsInterfaceNil return if there is no value under the interface
+func (scm *ShardCoordinatorMock) IsInterfaceNil() bool {
+	if scm == nil {
+		return true
+	}
+	return false
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply data/mock/shardCoordinatorMock.go]: Changed all to look like this.

File: data/mock/storerStub.go [hunk 18]
```diff
@@ -37,3 +37,11 @@ func (ss *StorerStub) ClearCache() {
 func (ss *StorerStub) DestroyUnit() error {
 	return ss.DestroyUnitCalled()
 }
+
+// IsInterfaceNil return if there is no value under the interface
+func (ss *StorerStub) IsInterfaceNil() bool {
+	if ss == nil {
+		return true
+	}
+	return false
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply data/mock/storerStub.go]: Changed all to look like this.

File: dataRetriever/mock/coordinatorStub.go [hunk 19]
```diff
@@ -31,3 +31,11 @@ func (coordinator *CoordinatorStub) SameShard(firstAddress, secondAddress state.
 func (coordinator *CoordinatorStub) CommunicationIdentifier(destShardID uint32) string {
 	return coordinator.CommunicationIdentifierCalled(destShardID)
 }
+
+// IsInterfaceNil return if there is no value under the interface
+func (coordinator *CoordinatorStub) IsInterfaceNil() bool {
+	if coordinator == nil {
+		return true
+	}
+	return false
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply dataRetriever/mock/coordinatorStub.go]: Changed all to look like this.

File: dataRetriever/requestHandlers/requestHandler.go [hunk 20]
```diff
@@ -184,3 +184,11 @@ func (rrh *ResolverRequestHandler) RequestHeaderByNonce(destShardID uint32, nonc
 		log.Debug(err.Error())
 	}
 }
+
+// IsInterfaceNil return if there is no value under the interface
+func (rrh *ResolverRequestHandler) IsInterfaceNil() bool {
+	if rrh == nil {
+		return true
+	}
+	return false
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply dataRetriever/requestHandlers/requestHandler.go]: Changed all to look like this.

File: facade/mock/apiResolverStub.go [hunk 21]
```diff
@@ -7,3 +7,11 @@ type ApiResolverStub struct {
 func (ars *ApiResolverStub) GetVmValue(address string, funcName string, argsBuff ...[]byte) ([]byte, error) {
 	return ars.GetVmValueHandler(address, funcName, argsBuff...)
 }
+
+// IsInterfaceNil returns if there is no value under the interface
+func (ars *ApiResolverStub) IsInterfaceNil() bool {
+	if ars == nil {
+		return true
+	}
+	return false
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply facade/mock/apiResolverStub.go]: Changed all to look like this.

File: process/factory/containers/vmContainer.go [hunk 22]
```diff
@@ -100,3 +100,11 @@ func (vmc *virtualMachinesContainer) Keys() [][]byte {
 	}
 	return keys
 }
+
+// IsInterfaceNil return if there is no value under the interface
+func (vmc *virtualMachinesContainer) IsInterfaceNil() bool {
+	if vmc == nil {
+		return true
+	}
+	return false
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply process/factory/containers/vmContainer.go]: Changed all to look like this.

File: node/node_test.go [hunk 23]
```diff
@@ -71,6 +71,9 @@ func getMessenger() *mock.MessengerStub {
 	return messenger
 }
 
+var marshalizer = &mock.MarshalizerMock{}
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply node/node_test.go]: created getMarshalizer and getHasher funcs.

File: process/smartContract/process.go (hunk 1 above)
Comment (BeniaminDrasovean) [reply process/smartContract/process.go]: done.

**Pull request merged**